### PR TITLE
Fix pip.list parsing of local packages

### LINKED ISSUE #58202: [BUG] salt.modules.pip:is_installed doesn't handle locally installed packages
**Description**

I'm setting up a fairly complex virtualenv with a few packages (e.g. cdecimal-2.3, internal packages, etc) that aren't published on pypi for whatever reason. I copy the package into the file system, check the hash and pip install it from a local file.  

The normal pip syntax for this is something like this:

     $ /var/venv/base/bin/pip install /var/venv/deps/cdecimal-2.3.tar.gz

But the same effect can be achieved through the `pip.installed` state:

```
/var/venv/deps/cdecimal-2.3.tar.gz
  pip.installed:
    - pip_bin: /var/venv/base/bin/pip
    - require:
      - virtualenv: /var/venv/base
      - file: /var/venv/deps/cdecimal-2.3.tar.gz
```

When you run `pip freeze --all` on this virtualenv, the output syntax of such packages is slightly different than the normal 'frozen' version number:

    $ /var/venv/base/bin/pip freeze --all

```
    ...
    cdecimal @ file:///var/venv/deps/cdecimal-2.3.tar.gz
    celery==3.1.26.post2
    ...
 ```

This syntax isn't handled by the pip.is_installed module:

```
    $ salt-call pip.is_installed cdecimal /var/venv/base
    [ERROR   ] Can't parse line 'cdecimal @ file:///var/venv/deps/cdecimal-2.3.tar.gz'
    local:
        False
```

The pip.is_installed module is used by other pip states to determine if a given library is installed. Hence, while it doesn't fail any states, it never sees that the locally installed package has been installed and installs it every time.

**Setup**
Please note that this setup is contrived for simplicity with a readily accessible example. There are other ways to install this particular package using `--find-links` that works around the problem.  For self-maintained dist packages, I haven't really found a good solution that's as straightforward as simply pip installing the dist tarball directly from the file system.

```
/var/venv/base:
  virtualenv.managed: []
/var/venv/deps/cdecimal-2.3.tar.gz:
  file.managed:
    - source: https://www.bytereef.org/software/mpdecimal/releases/cdecimal-2.3.tar.gz
    - source_hash: d737cbe43ed1f6ad9874fb86c3db1e9bbe20c0c750868fde5be3f379ade83d8b
    - makedirs: True
    - creates: /var/venv/deps/cdecimal-2.3.tar.gz
  pip.installed:
    - pip_bin: /var/venv/base/bin/pip
    - require:
      - virtualenv: /var/venv/base
      - file: /var/venv/deps/cdecimal-2.3.tar.gz
```

**Steps to Reproduce the behavior**

```
    $ salt-call pip.is_installed cdecimal /var/venv/base
    [ERROR   ] Can't parse line 'cdecimal @ file:///var/venv/deps/cdecimal-2.3.tar.gz'
    local:
        False
```

**Expected behavior**
For completeness, `pip.is_installed` could/should parse the `pip freeze` syntax of "[pkgname] @ file:///..." to determine that a locally installed package is present.

I think it would be as simple as splitting on the first '@', trimming spaces and comparing the first element to the specified package name.

**Screenshots**
N/A

**Versions Report**
<details><summary>salt --versions-report</summary>

```
# salt --versions-report
Salt Version:
           Salt: 3001.1

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: 4.3.0
          gitdb: 2.0.3
      gitpython: 2.1.8
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: 3.4.7
         pygit2: Not Installed
         Python: 3.6.9 (default, Jul 17 2020, 12:50:27)
   python-gnupg: 0.4.1
         PyYAML: 3.12
          PyZMQ: 17.1.2
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5

System Versions:
           dist: ubuntu 18.04 Bionic Beaver
         locale: UTF-8
        machine: x86_64
        release: 5.3.0-1032-aws
         system: Linux
        version: Ubuntu 18.04 Bionic Beaver

```
</details>

**Additional context**
N/A

### LINKED ISSUE #60557: [BUG] Regression: pip.installed does not pass env_vars when calling pip.list to check if package is already installed
**Description**
This is a regression of #46127 and was introduced a few months after the original fix merged

See that bug for more details about how to reproduce the issue

**Additional context**

[This commit](https://github.com/saltstack/salt/commit/4fec8f6bcce7965b5294939b906d0e344294c651) made a change to cache the installed list using `pip.list` to speed up pip execution. However, the call to pip list does not pass `env_vars` [(see 3003.1 code here)](https://github.com/saltstack/salt/blob/2ecd867850f85e2ca68a30838d44efafcaddc9ea/salt/states/pip_state.py#L843) and hence the bug has reappeared.

```
pip_list = __salt__["pip.list"](bin_env=bin_env, user=user, cwd=cwd)
```

Fix is trivial by passing `env_vars=env_vars` to this function.

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
See issues for details.

### What issues does this PR fix or reference?
Fixes: #58202
Fixes: #60557

### Previous Behavior
The previous functionality parsed `pip freeze` output line by line as strings.

### New Behavior
The new functionality uses `pip list` with JSON output, which allows Salt to parse as a data structure which already has components such as name and version separated.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by nicholasmhughes** [comment]
re-run pr-arch-lts-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-windows-2019-x64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-ubuntu-2004-arm64-py3-pytest

File: salt/modules/pip.py [hunk 1]
```diff
@@ -1301,6 +1305,73 @@ def list_(prefix=None, bin_env=None, user=None, cwd=None, env_vars=None, **kwarg
...
+
+        salt '*' pip.list salt
+    """
+
+    packages = {}
+    cwd = _pip_bin_env(cwd, bin_env)
+    cur_version = version(bin_env, cwd, user=user)
+
+    # Pip started supporting the ability to output json starting with 9.0.0
+    min_version = "9.0"
```
Comment (s0undt3ch) [inline salt/modules/pip.py:1333]: I don't believe any of the platforms we currently support ships pip version 9.

Maybe we should just error out if it's that old?

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
And, if we drop support for older pip, we don't really need a separate list function(`list_freeze_parse`) doe we?

Anyway, let see what the rest of the team thinks about dropping really old pip versions.

File: salt/modules/pip.py (hunk 1 above)
Comment (nicholasmhughes) [reply salt/modules/pip.py:1333]: Any update if we can merge for 3006 as-is and then drop support for old versions in 3007?

**Comment by Ch3LL** [comment]
@s0undt3ch can you re-review please

File: salt/modules/pip.py (hunk 1 above)
Comment (waynew) [reply salt/modules/pip.py:1331]: Per Open Hour conversation today, it's probably fine to go ahead and leave this for 3006. For 3007 we'll want to go ahead and drop support for pip 9 or less :+1: (should probably convert this comment to an issue)

Comment (nicholasmhughes) [reply salt/modules/pip.py:1331]: talked about this in Open Hour. deferring to 3007

Comment (nicholasmhughes) [reply salt/modules/pip.py:1331]: opened #63524

**Review by Ch3LL — APPROVED** [review]

**Review by garethgreenaway — APPROVED** [review]

**Review by twangboy — APPROVED** [review]

**Comment by nicholasmhughes** [comment]
re-run pr-photon-3-x86_64-py3-pytest

**Pull request merged**